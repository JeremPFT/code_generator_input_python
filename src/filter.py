class Filter():
    # trying to build a OOP filters family

    registered = {}

    def __init__(self, tag_name):
        self.tag_name = tag_name.lower()

    def process(self, tag_value):
        the_process = Filter.PROCESSING[self.index]
        return the_process(self, tag_value)

    def register(Filter_Class):
        registered[Filter_Class, Filter_Class.tag_name]


class Filter_Capitalize(Filter):
    def __init__(self):
        super().__init__("capitalize")
        Filter.register(self.__class__, self.tag_name)

    def process(self, tag_value):
        words = tag_value.split("_")
        splitted_image = []
        for word in words:
            splitted_image.append(word.capitalize())
        return "_".join(splitted_image)


class Filter_Upper(Filter):
    def __init__(self):
        super().__init__("capitalize")
        Filter.registered.append(self)

    def process(self, tag_value):
        return tag_value.upper()


class Filter_Double(Filter):
    def __init__(self):
        super().__init__("capitalize")
        Filter.registered.append(self)

    def process(self, tag_value):
        result = ""
        for car in argument:
            result += car
            result += car
        return result
