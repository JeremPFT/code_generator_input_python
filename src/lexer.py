import ply.lex as lex

reserved = {
    'abstract'         : 'ABSTRACT',
    'and'              : 'AND',
    'command'          : 'COMMAND',
    'end'              : 'END',
    'exceptions'       : 'EXCEPTIONS',
    'field'            : 'FIELD',
    'implementation'   : 'IMPLEMENTATION',
    'in'               : 'IN',
    'inout'            : 'INOUT',
    'is'               : 'IS',
    'limited'          : 'LIMITED',
    'operation'        : 'OPERATION',
    'or'               : 'OR',
    'out'              : 'OUT',
    'output_directory' : 'OUTPUT_DIRECTORY',
    'package'          : 'PACKAGE',
    'post'             : 'POST',
    'pre'              : 'PRE',
    'project'          : 'PROJECT',
    'query'            : 'QUERY',
    'readme_brief'     : 'README_BRIEF',
    'readme_title'     : 'README_TITLE',
    'return'           : 'RETURN',
    'subprogram'       : 'SUBPROGRAM',
    'type'             : 'TYPE',
    'use'              : 'USE',
    'value_object'     : 'VALUE_OBJECT',
    'vector'           : 'VECTOR',
    'with'             : 'WITH',
}

tokens = [
    'DOUBLEPERIOD', # ! before period
    'COLONEQ', # ! before colon, before equal
] + [
    'STAR',
    'LPAREN',
    'RPAREN',
    'SEMICOLON',
    'PERIOD',
    'COLON',
    'AMP',
    'IDENTIFIER',
    'COMMENT',
    'STRING_VALUE',
    'INTEGER_VALUE',
    'SUPERIOR',
    'INFERIOR',
    'EQUAL',
    'VALUE',
] + list(reserved.values())

t_ignore  = ' \t'

def t_STAR(t):
    r'\*'
    return t

def t_LPAREN(t):
    r'\('
    return t

def t_DOUBLEPERIOD(t):
    r'\.\.'
    return t

def t_RPAREN(t):
    r'\)'
    return t

def t_SEMICOLON(t):
    r';'
    return t

def t_COLON(t):
    r':'
    return t

def t_COLONEQ(t):
    r':='
    return t

def t_AMP(t):
    r'&'
    return t

def t_SUPERIOR(t):
    r'>'
    return t

def t_INFERIOR(t):
    r'<'
    return t

def t_EQUAL(t):
    r'='
    return t

def t_IDENTIFIER(t):
    r'[a-zA-Z][a-zA-Z_0-9.]*'
    t.type = reserved.get(t.value, 'IDENTIFIER')
    return t

def t_COMMENT(t):
    r'--.*'
    pass

def t_STRING_VALUE(t):
    r'".*?"'
    return t

def t_INTEGER_VALUE(t):
    r'[0-9]+'
    return t

def t_VALUE(t):
    r'(".*")|([a-z0-9]+)'

def t_newline(t):
    r'\n+'
    t.lexer.lineno += len(t.value)

def t_eof(t):
    t.lexer.lineno = 1
    return None

def t_error(t):
    print("Illegal character '%s'" % t.value[0])
    t.lexer.skip(1)

def find_column(input, token):
    line_start = input.rfind('\n', 0, token.lexpos) + 1
    print("line_start = {!s}, lexpos = {!s}".format(line_start, token.lexpos))
    return (token.lexpos - line_start) + 1

lexer = lex.lex(debug = False)
lexer.lineno = 1

# def build_lexer():
#     return lex.lex(debug = False)

def test_lexer(data):
    lexer.input(data)

    token = lexer.token()
    while token != None :
        print(token)
        print('column: ' + str(find_column(data, token)))
        token = lexer.token()

if __name__ == '__main__':
    test_lexer()
