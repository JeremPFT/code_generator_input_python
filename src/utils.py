import os
import shutil

def dbg(str):
    print(str)


class indent:
    shift = 2
    value = 0

    @staticmethod
    def str():
        return indent.value * " "

    @staticmethod
    def incr():
        indent.value += indent.shift

    @staticmethod
    def decr():
        indent.value -= indent.shift


def capitalize_identifier(identifier):
    orig_list = identifier.split("_")
    result_list = []
    for str in orig_list:
        result_list.append(str.capitalize())
    return '_'.join(result_list)


def directory(path):
    'normalize given path and replace ~/ with HOME path'

    result = os.path.normpath(path)
    if result.startswith("~\\"):
        home = os.path.normpath(os.environ["HOME"])
        result = os.path.join(home, result[2:])
    return result

def rmdir(path):
    shutil.rmtree(directory(path))

def build_dir(dir):
    if not os.path.exists(dir):
        os.makedirs(dir)

def dir_exists(path, name):
    path = os.path.normpath(path)
    if not os.path.exists(path):
        print("ERROR: %s doesn't exist : %s" % (name, path))
        return False
    else:
        print("path %s exists : %s" % (name, path))
        return True

def text_files_match(expected, obtained):
    '''
    compare each line of given texts
    '''

    dbg("%s expected: %s, obtained: %s" % (text_files_match.__name__,
                                           expected,
                                           obtained))

    expected_lines = open(directory(expected), "r").read().split('\n')
    obtained_lines = open(directory(obtained), "r").read().split('\n')

    error_msg = ""
    for line_num in range(0, len(expected_lines)):
        expected_line = expected_lines[line_num]
        obtained_line = obtained_lines[line_num]
        if expected_line != obtained_line:
            error_msg = 'ERROR line ' + str(line_num) + '\n' \
                + 'expects "' + expected_line + '"\n' \
                + 'got     "' + obtained_line + '"'
            print(error_msg)

    if error_msg == "":
        print("matching files\n%s\n%s" % (expected, obtained))
        return True
    else:
        return False
