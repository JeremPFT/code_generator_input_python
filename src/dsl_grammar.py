#! python3

import os
import ply.yacc as yacc

from src.lexer import tokens

from src.uml_model import *

from src.utils import dbg

'''
(fset 'copy-grammar-rule
[?\C-s ?: ?\C-j ?  ?\' ?\' ?\' return ?\M-f ?\M-b ?\C-  ?\C-s ?\' ?\' ?\' ?\C-a ?\M-w ?\C-x ?o ?\C-y ?\C-x ?o])
'''

'''
project_item                       : project_init project_content project_close
project_init                       : PROJECT IDENTIFIER
project_content                    : output_directory project_type package_list
project_close                      : END PROJECT IDENTIFIER SEMICOLON
project_close                      : END PROJECT SEMICOLON
package_list                       :
package_list                       : package_list package
package                            : package_init package_content package_close
package_init                       : PACKAGE IDENTIFIER
package_content                    : dependance_list packageable_element_list
package_close                      : END PACKAGE IDENTIFIER SEMICOLON
package_close                      : END PACKAGE SEMICOLON
dependance_list                    :
dependance_list                    : dependance_list dependance
dependance                         : WITH IDENTIFIER
dependance                         : USE IDENTIFIER
dependance                         : LIMITED WITH IDENTIFIER
packageable_element_list           :
packageable_element_list           : packageable_element_list packageable_element
packageable_element                : operation
                                   | type_item
type_item                          : exception_block                 // TODO
                                   | value_object
value_object                       : value_object_init value_object_content value_object_close
value_object_init                  : value_object_init_abstract_inherit
                                   | value_object_init_abstract
                                   | value_object_init_inherit
                                   | value_object_init_simple
value_object_init_abstract_inherit : ABSTRACT VALUE_OBJECT IDENTIFIER LPAREN IDENTIFIER RPAREN
value_object_init_abstract         : ABSTRACT VALUE_OBJECT IDENTIFIER
value_object_init_inherit          : VALUE_OBJECT IDENTIFIER LPAREN IDENTIFIER RPAREN
value_object_init_simple           : VALUE_OBJECT IDENTIFIER
value_object_content               : dependance_list feature_list
value_object_close                 : END VALUE_OBJECT IDENTIFIER SEMICOLON
value_object_close                 : END VALUE_OBJECT SEMICOLON
feature_list                       :
feature_list                       : feature_list feature
feature                            : property SEMICOLON
                                   | operation SEMICOLON
operation                          : operation_init parameter_list operation_return
operation_init                     : OPERATION IDENTIFIER
operation_return                   :
operation_return                   : RETURN IDENTIFIER
parameter_list                     :
parameter_list                     : LPAREN parameter_item RPAREN
parameter_list                     : LPAREN parameter_item_list parameter_item RPAREN
parameter_item_list                : parameter_item SEMICOLON
parameter_item_list                : parameter_item_list parameter_item
parameter_item                     : IDENTIFIER COLON direction IDENTIFIER
direction                          : INOUT
                                   | OUT
                                   | IN
property                           : IDENTIFIER COLON IDENTIFIER
project_type                       : TYPE IDENTIFIER SEMICOLON
output_directory                   : OUTPUT_DIRECTORY string
string                             : string AMP STRING_VALUE
string                             : STRING_VALUE
'''

def check_closing_name(p, class_name):
    if p[3] != p[-2].name:
        print("WARNING closed %s %s is not opened %s %s" % \
              (class_name, p[3], class_name, p[-2].name))

#
# - project
#

def p_project_item_no_named(p):
    '''
    project : PROJECT OUTPUT_DIRECTORY
    '''
    raise NoNameError("SyntaxError line {!s}: project need a name".format(p.lineno(1)))

def p_project_item(p):
    '''
    project : project_init project_content project_close
    '''
    p[0] = p[1]

def p_project_init(p):
    '''
    project_init : PROJECT IDENTIFIER
    '''
    p[0] = Project(name = p[2])

# def p_project_init_no_name(p):
#     '''
#     project_init : PROJECT
#     '''
#     # p.parser.error = NoNameError()
#     # print("SyntaxError line {!s}: project need a name".format(p.lineno(1)))
#     # p[0] = Project(name = "")
#     p.parser.error = NoNameError("SyntaxError line {!s}: project need a name".format(p.lineno(1)))
#     p_error(p)

def p_project_content(p):
    '''
    project_content : output_directory project_type readme_content package_list
    '''
    p[-1].set_output_directory(p[1])
    p[-1].set_type(p[2])

    for package_item in p[3]:
        p[-1].add_package(package_item)

def p_readme_content(p):
    '''
    reamde_content : REAMDE_TITLE string REAMDE_BRIEF string
    '''
    p[-2].set_readme_title(p[2])
    p[-2].set_readme_brief(p[4])

def p_project_close_with_name(p):
    '''
    project_close : END PROJECT IDENTIFIER SEMICOLON
    '''
    check_closing_name(p, "project")

def p_project_close(p):
    '''
    project_close : END PROJECT SEMICOLON
    '''
    pass

#
# - package
#

def p_package_list_empty(p):
    '''
    package_list :
    '''
    p[0] = []

def p_package_list_more(p):
    '''
    package_list : package_list package
    '''
    p[1].append(p[2])
    p[0] = p[1]

def p_package_item(p):
    '''
    package : package_init package_content package_close
    '''
    p[0]                     = p[1]
    p.parser.current_package = p[0]

def p_package_init(p):
    '''
    package_init : PACKAGE IDENTIFIER
    '''
    p[0] = Package(name = p[2])
    p.parser.current_package = p[0]

def p_package_content(p):
    '''
    package_content : dependance_list packageable_element_list
    '''
    p[-1].dependance_list          = p[1]
    p[-1].packageable_element_list = p[2]

def p_package_close_with_name(p):
    '''
    package_close : END PACKAGE IDENTIFIER SEMICOLON
    '''
    check_closing_name(p, "package")

def p_package_close(p):
    '''
    package_close : END PACKAGE SEMICOLON
    '''
    pass

#
# - dependancy
#

def p_dependance_list_empty(p):
    '''
    dependance_list :
    '''
    p[0] = []

def p_dependance_list_more(p):
    '''
    dependance_list : dependance_list dependance
    '''
    p[1].append(p[2])
    p[0] = p[1]

def p_dependance_with(p):
    '''
    dependance : WITH IDENTIFIER
    '''
    p[0] = Dependance(mode = p[1], imported_unit = p[2])

def p_dependance_use(p):
    '''
    dependance : USE IDENTIFIER
    '''
    p[0] = Dependance(mode = p[1], imported_unit = p[2])

def p_dependance_limited_with(p):
    '''
    dependance : LIMITED WITH IDENTIFIER
    '''
    p[0] = Dependance(mode = p[1] + p[2], imported_unit = p[3])

#
# - packageable_element
#

def p_packageable_element_list_empty(p):
    '''
    packageable_element_list :
    '''
    p[0] = []

def p_packageable_element_list_more(p):
    '''
    packageable_element_list : packageable_element_list packageable_element
    '''
    p[1].append(p[2])
    p[0] = p[1]

def p_packageable_element_item(p):
    '''
    packageable_element : operation
                        | type_item
    '''
    p[0] = p[1]

    p.parser.current_package.add_owned_member(p[0])

def p_type_item(p):
    '''
    type_item : value_object
              | exception_block
    '''
    p[0] = p[1]

def p_exception_block(p):
    '''
    exception_block : EXCEPTIONS exception_list exception_item END EXCEPTIONS SEMICOLON
    '''

def p_exception_list_empty(p):
    '''
    exception_list :
    '''
    p[0] = []

def p_exception_list_more(p):
    '''
    exception_list : exception_list exception_item
    '''
    p[1].append(p[2])
    p[0] = [1]

def p_exception_item(p):
    '''
    exception_item : IDENTIFIER
    '''
    p[0] = p[1]

#
# - value_object
#

def p_value_object_item(p):
    '''
    value_object : value_object_init value_object_content value_object_close
    '''
    p[0] = p[1]

def p_value_object_init(p):
    '''
    value_object_init : value_object_init_abstract_inherit
                      | value_object_init_abstract
                      | value_object_init_inherit
                      | value_object_init_simple
    '''
    p[0] = p[1]
    parser.current_class = p[0]

def p_value_object_init_abstract_inherit(p):
    '''
    value_object_init_abstract_inherit : ABSTRACT VALUE_OBJECT IDENTIFIER LPAREN IDENTIFIER RPAREN
    '''
    is_abstract = True
    parent_name = p[5]
    name        = p[3]

    p[0] = Class(name        = name,
                 owner       = p.parser.current_package,
                 is_abstract = is_abstract,
                 parent_name = parent_name)

def p_value_object_init_abstract(p):
    '''
    value_object_init_abstract : ABSTRACT VALUE_OBJECT IDENTIFIER
    '''
    is_abstract = True
    name        = p[3]

    p[0] = Class(name        = name,
                 owner       = p.parser.current_package,
                 is_abstract = is_abstract)

def p_value_object_init_inherit(p):
    '''
    value_object_init_inherit : VALUE_OBJECT IDENTIFIER LPAREN IDENTIFIER RPAREN
    '''
    parent_name = p[4]
    name        = p[2]

    p[0] = Class(name        = name,
                 owner       = p.parser.current_package,
                 parent_name = parent_name)

def p_value_object_init_simple(p):
    '''
    value_object_init_simple : VALUE_OBJECT IDENTIFIER
    '''
    name = p[2]
    p[0] = Class(name  = name,
                 owner = p.parser.current_package)

def p_value_object_content(p):
    '''
    value_object_content : dependance_list feature_list
    '''
    p[-1].dependance_list = p[1]

    for feature in p[2]:
        if feature.__class__.__name__ == Property.__name__:
            p[-1].add_property(feature)
        elif feature.__class__.__name__ == Operation.__name__:
            p[-1].add_operation(feature)

def p_value_object_close_with_name(p):
    '''
    value_object_close : END VALUE_OBJECT IDENTIFIER SEMICOLON
    '''
    check_closing_name(p, "value_object")

def p_value_object_close(p):
    '''
    value_object_close : END VALUE_OBJECT SEMICOLON
    '''
    pass

#
# - feature
#

def p_feature_list_none(p):
    '''
    feature_list :
    '''
    p[0] = []

def p_feature_list_more(p):
    '''
    feature_list : feature_list feature
    '''
    p[1].append(p[2])
    p[0] = p[1]

def p_feature_item(p):
    '''
    feature : property SEMICOLON
            | operation SEMICOLON
    '''
    p[0] = p[1]

#
# - operation
#

def p_operation_item(p):
    '''
    operation : operation_init parameter_list operation_return
    '''
    operation     = p[1]
    parameters    = p[2]
    returned_type = p[3]

    for parameter in parameters:
        operation.add_parameter(parameter)

    if returned_type != None:
        operation.add_parameter(returned_type)

    p[0] = operation

def p_operation_init(p):
    '''
    operation_init : OPERATION IDENTIFIER
    '''
    p[0] = Operation(name = p[2])

def p_operation_return_none(p):
    '''
    operation_return :
    '''
    p[0] = None

def p_operation_return_one(p):
    '''
    operation_return : RETURN IDENTIFIER
    '''
    p[0] = Parameter(name      = "result",
                     of_type   = p[2],
                     direction = Parameter.DIRECTION_RETURN)

def p_parameter_list(p):
    '''
    parameter_list :
    '''
    p[0] = []

def p_parameter_list_one(p):
    '''
    parameter_list : LPAREN parameter_item RPAREN
    '''
    p[0] = [p[2]]

def p_parameter_list_more(p):
    '''
    parameter_list : LPAREN parameter_item_list parameter_item RPAREN
    '''
    p[0] = p[2]
    p[0].append(p[3])

def p_parameter_item_list_one(p):
    '''
    parameter_item_list : parameter_item SEMICOLON
    '''
    p[0] = [p[1]]

def p_parameter_item_list_more(p):
    '''
    parameter_item_list : parameter_item_list parameter_item
    '''
    p[0] = p[1]
    p[0].append(p[2])

def p_parameter_item(p):
    '''
    parameter_item : IDENTIFIER COLON direction IDENTIFIER
    '''
    p[0] = Parameter(name = p[1], direction = p[3], of_type = p[4])

def p_direction(p):
    '''
    direction : INOUT
              | OUT
              | IN
    '''
    p[0] = p[1]

#
# - implementation
#

# pseudocode def p_implementation_from_file(p):
# pseudocode     '''
# pseudocode     implementation : IMPLEMENTATION string SEMICOLON
# pseudocode     '''
# pseudocode
# pseudocode     implementation_file = open(p[2])
# pseudocode
# pseudocode     # TODO following is pseudo code
# pseudocode
# pseudocode     declaration = substring(path = implementation_file,
# pseudocode                             from = position_of ("[^a-z_]is[^a-z_]"),
# pseudocode                             end  = position_of ("[^a-z_]begin[^a-z_]"))
# pseudocode
# pseudocode     body = substring(path = implementation_file,
# pseudocode                      from = position_of ("[^a-z_]begin[^a-z_]"),
# pseudocode                      end  = last_position_of ("[^a-z_]end[^a-z_]"))
# pseudocode
# pseudocode def p_implementation(p):
# pseudocode     '''
# pseudocode     implementation : declaration_item body_item
# pseudocode     '''
# pseudocode     pass
# pseudocode
# pseudocode def p_declaration(p):
# pseudocode     '''
# pseudocode     declaration_item : DECLARATION declaration_content
# pseudocode     '''

#
# - property
#

def p_property_item(p):
    '''
    property : IDENTIFIER COLON IDENTIFIER
    '''
    p[0] = Property(name = p[1], of_type = p[3])

#
# - others
#

def p_project_type(p):
    '''
    project_type : TYPE IDENTIFIER SEMICOLON
    '''
    prj_type = p[2]

    if not prj_type in Project.TYPES:
        print("!! SyntaxError: project type {!r} undefined line {}".format(prj_type, p.lineno(1)))
        raise SyntaxError

    p[0] = p[2]

def p_output_directory(p):
    '''
    output_directory : OUTPUT_DIRECTORY string SEMICOLON
    '''
    p[0] = p[2]

def p_string_one_or_more(p):
    '''
    string : string AMP STRING_VALUE
    '''
    left  = p[1].replace('"', '')
    right = p[3].replace('"', '')
    p[0]  = '"' + left + right + '"'

def p_string_one(p):
    '''
    string : STRING_VALUE
    '''
    p[0] = p[1]

def p_error(p):
    if p == None:
        print("!! SyntaxError: end of file")
        return

    if p.parser.error != None:
        print(str(p.parser.error))

        tok = None
        while True:
            tok = parser.token()
            if tok == None:
                break
            else:
                print("ignoring token line {!s}".format(tok.lineno))
    else:
        print("!! SyntaxError in input")
        message = "line %s: unexpected %s %s" % (str(p.lineno), str(p.type), str(p))
        print(message)

parser = yacc.yacc(debug = False)
parser.error = None
parser.current_project   = None
parser.current_package   = None
parser.current_class     = None
parser.current_procedure = None

def test_grammar(data):
    result = parser.parse(data, debug = False)
    print("return object " + str(type(result)))
    return result

if __name__ == '__main__':
    prj = test_grammar()
    if prj != None:
        print(os.linesep + ("=" * 60) + os.linesep)
        print(prj)
        print(os.linesep + "=" * 60)
