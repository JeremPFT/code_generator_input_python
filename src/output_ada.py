import os

from jinja2 import (
    Environment,
    FileSystemLoader,
    PackageLoader,
)
# https://palletsprojects.com/p/jinja/
#
# from jinja2 import Template

from src.uml_model import *
from src.utils import (indent, dbg, capitalize_identifier, build_dir)

class Output_Ada():
    def __init__(self):
        self.env = Environment(loader      = FileSystemLoader('templates'),
                               trim_blocks = True)

    def output(self, project):
        self._output_project(project)

    def _output_project(self, project):
        template = self.env.get_template('lib_project.gpr')

        print()
        prj_dir = os.path.normpath(project.output_directory())
        dbg("buildind directory " + prj_dir)

        print("build dir '%s'" % (prj_dir))
        build_dir(prj_dir)
        os.chdir(prj_dir)
        print("build dir '%s'" % ("src"))
        build_dir("src")

        rendered_template = template.render(project_name = capitalize_identifier(project.name))

        f = open(project.name + ".gpr", "w")
        f.write(rendered_template)
        f.close()

        print("file generated: ")
        print(rendered_template)
