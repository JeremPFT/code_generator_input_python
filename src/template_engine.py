#! python3

''' A simple template engine which should follow ada library Template Parser.

See http://docs.adacore.com/live/wave/aws/html/template_parser
'''

import os

from src.utils import (
    text_files_match,
    )

class Statement():
    def __init__(self, keyword, stmt_start, stmt_stop):
        self.start = stmt_start
        self.stop = stmt_stop
        self.keyword = keyword

    def render(self, text):
        return text


class Filter():
    def __init__(self, keyword):
        lower_keyword = keyword.lower()

        if not lower_keyword in Filter.KEYWORDS:
            raise ValueError("unknown filter '%s'" % (keyword))

        self.index = Filter.KEYWORDS.index(lower_keyword)

    def process(self, argument):
        the_process = Filter.PROCESSING[self.index]
        return the_process(self, argument)

    def __do_capitalize(self, argument):
        words = argument.split("_")
        splitted_image = []
        for word in words:
            splitted_image.append(word.capitalize())
        return "_".join(splitted_image)

    def __do_upper(self, argument):
        return argument.upper()

    def __do_double(self, argument):
        image = ""
        for car in argument:
            image += car
            image += car
        return image

    KEYWORDS = (
        "capitalize",
        "upper",
        "double",
    )

    PROCESSING = (
        __do_capitalize,
        __do_upper,
        __do_double,
    )


class Template_Engine():
    def __init__(self, template_directory):
        self.tag_start = "@_"
        self.tag_stop = "_@"
        self.comment = "@@--"
        # to get a behavior like ada library template_parser

        self.__template_directory = ""
        self._set_template_directory(template_directory)

        self.dico = None

    def _set_template_directory(self, path):
        if type(path) != str:
            raise TypeError("template_directory has to be a string")

        if not os.path.exists(path):
            raise ValueError("template_directory not found: '%s'" % (path))

        self.__template_directory = path

    def _render_tag(self, string):
        string = string[len(self.tag_start):-len(self.tag_stop)]
        substrings = string.split(":")

        rendered = substrings[-1].lower()

        for key, value in self.dico.items():
            lower_key = key.lower()
            lower_value = value.lower()
            if lower_key in rendered:
                rendered = rendered.replace(lower_key, lower_value)
                break

        if len(substrings) > 1:
            for idx in range(len(substrings) - 2, -1, -1):
                filter = Filter(substrings[idx])
                rendered = filter.process(rendered)

        return rendered

    def render(self, file_name, dico):
        self.dico = dico

        file_full_path = os.path.join(self.__template_directory, file_name)

        template = open(file_full_path, "r")
        count    = 0
        rendered = []

        for line in template:
            count += 1

            if line.strip().startswith(self.comment):
                continue

            start_index = line.find(self.tag_start)
            stop_index  = line.rfind(self.tag_stop)
            shift       = len(self.tag_stop)

            if start_index != -1 and stop_index != -1:
                rendered.append(line[:start_index]
                                + self._render_tag(line[start_index:stop_index + shift])
                                + line[stop_index + shift:])

            elif start_index != -1 and stop_index == -1:
                print("WARNING line %s, tag not close :" + line[start_index:] )

            elif start_index == -1 and stop_index != -1:
                print("WARNING line %s, closing tag not open")

            else:
                rendered.append(line)

        count = 0

        template.close()

        return ''.join(rendered)

def test_engine():
    template_dir = "d:/Users/jpiffret/AppData/Roaming/Dropbox/projets_perso/ada/code_generator_py/templates"
    te = Template_Engine(template_dir)

    os.chdir(template_dir)

    output = te.render("lib_project.gpr", {
        "Project_Name" : "essai_1_azerty",
        "A" : "zerty",
        "b" : "ercredi",
    })

    result = open("lib_project.out", "w")
    result.write(output)
    result.close()
    text_files_match(expected = "lib_project.in",
                     actual   = "lib_project.out")

if __name__ == '__main__':
    test_engine()
