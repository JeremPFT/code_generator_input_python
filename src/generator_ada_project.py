import os

from src.uml_model import *
from src.template_engine import Template_Engine
from src.utils import (
    indent,
    dbg,
    capitalize_identifier,
    build_dir,
    directory,
)


class Generator_Ada_Project():
    def __init__(self, template_engine):
        if template_engine.__class__ != Template_Engine:
            raise TypeError("template_engine must be of Template_Engine class")
        self.__template_engine = template_engine
        self.__project = None

    def output(self, project):
        print(str(project))

        assert project != None and type(project) == Project, \
            "not a Project instance: " + str(type(project))

        if project == None:
            print("project is null, error during parsing")
            return

        self.__project = project
        self.__output_project()

    def __create_tree(self):
        prj_dir = directory(self.__project.output_directory)
        build_dir(prj_dir)
        build_dir(os.path.join(prj_dir, "src"))

    def __output_project(self):
        self.__create_tree()

        cap_project_name = capitalize_identifier(self.__project.name)

        te = self.__template_engine

        dico = {'project_name' : cap_project_name}
        tmpl_name = "project.gpr"
        rendered_template = te.render (file_name = tmpl_name,
                                       dico = dico)

        prj_dir = directory(self.__project.output_directory)
        gpr_file_name = self.__project.name + ".gpr"
        gpr_file_name = os.path.join(prj_dir, gpr_file_name)
        f = open(gpr_file_name, "w")
        f.write(rendered_template)
        f.close()
