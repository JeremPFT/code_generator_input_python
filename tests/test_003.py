import unittest

from src.generator_ada_project import Generator_Ada_Project
from src.dsl_grammar import parse_input
from src.template_engine import Template_Engine

from src.utils import (
    directory,
    rmdir,
    dir_exists,
    text_files_match,
)


class Test_Project_Nominal(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        print(__class__)
        filename = "~/workspace/code_generator_py/tests/test_003/test_003.dsl"
        cls.project = parse_input(filename)

        template_dir = directory("~/workspace/code_generator_py/tests/test_003")
        template_engine = Template_Engine(template_directory = template_dir)
        generator = Generator_Ada_Project(template_engine)
        generator.output(cls.project)

    @classmethod
    def tearDownClass(cls):
        rmdir("~/tests")

    def test_field_name(self):
        print(__name__)
        self.assertEqual(Test_Project_Nominal.project.name,
                         "test_003",
                         "project name")

    def test_field_output_directory(self):
        self.assertEqual(Test_Project_Nominal.project._output_directory,
                         directory("~/tests/test_003"),
                         "output directory")

    def test_field_title(self):
        self.assertEqual(Test_Project_Nominal.project._title,
                         "test 003 title",
                         "project title")

    def test_field_brief(self):
        self.assertEqual(Test_Project_Nominal.project._brief,
                         "test 003 brief",
                         "project brief")

    def _setdown(self):
        super()._setdown()

    def _build_test_list(self):
        for test in (
                check_project,
                check_prj_dir,
                check_src_dir,
                check_gpr_file,
        ):
            self._add_test(test)

    def get_project(self):
        return self.__project
