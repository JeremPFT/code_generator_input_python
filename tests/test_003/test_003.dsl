project test_003

  output_directory "~/tests/test_003";

  type static_library;

  readme_title "test 003 title"
  readme_brief "test 003" & " brief"

  package a_first_package
    value_object element is abstract
      property comment : [0-n] ordered
    end value_object;
  end package a_first_package;

end project test_003;
