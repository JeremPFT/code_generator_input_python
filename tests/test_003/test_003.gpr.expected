with "../common/shared.gpr";

library project Test_003 is

  for Create_Missing_Dirs use "True";

  Src_Lst := ();
  Src_Lst := Src_Lst & "./src_lib";
  for Source_Dirs use Src_Lst;

  Exc_Src_Lst := ();
  for Excluded_Source_Files use Exc_Src_Lst;

  for Object_Dir use Shared.Object_Dir;
  for Library_Dir use Shared.Library_Dir;

  for Library_Name use "test_003";
  for Library_Kind use "static";

  package Compiler renames Shared.Compiler;

  package Builder renames Shared.Builder;

end Test_003;
